### PR TITLE
mchprs: init at 0.4.1, nixos/mchprs: init module

### PR DESCRIPTION
**Pull request by gabriel-doriath-dohler**
###### Description of changes

[MCHPRS](https://github.com/MCHPR/MCHPRS) is multithreaded Minecraft server built for redstone.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/games/mchprs/default.nix [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+, rustPlatform
+, fetchFromGitHub
+, pkg-config
+, sqlite
+, openssl
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "mchprs";
+  version = "unstable";
```
Comment (RaitoBezarius) [inline pkgs/games/mchprs/default.nix:11]:
```suggestion
  version = "unstable-2023-07-01";
```

File: pkgs/games/mchprs/default.nix [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+    outputHashes = {
+      "hematite-nbt-0.5.2" = "sha256-knBmH/32JJclhFZbKTNx5XgLSQ2rIrXUGu8uoAHAXMk=";
+    };
+  };
+
+  # Rust nightly features are needed for the `pattern` feature of the dependency `regex`
+  RUSTC_BOOTSTRAP = true;
```
Comment (RaitoBezarius) [inline pkgs/games/mchprs/default.nix:28]:
```suggestion
  RUSTC_BOOTSTRAP = 1;
```

**Review by JulienMalka — COMMENTED** [review]
Built correctly on aarch64-linux.

**Review by RaitoBezarius — APPROVED** [review]

**Comment by RaitoBezarius** [comment]
Result of `nixpkgs-review pr 240913` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mchprs</li>
  </ul>
</details>

**Comment by gabriel-doriath-dohler** [comment]
Tested basic functionality on x86_64-linux.
No options for luckperms are provided. It could be a good idea to handle the password for the luckperms DB.
NixOS tests are difficult because no equivalent of mcrcon exist for MCHPRS.
The module contains many options, but they are required for starting the service.
Maybe further hardening is possible? Idk.

**Comment by RaitoBezarius** [comment]
Result of `nixpkgs-review pr 240913` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mchprs</li>
  </ul>
</details>

File: nixos/modules/services/games/mchprs.nix [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+    };
+    users.groups.mchprs = { };
+
+    systemd.services.mchprs = {
+      description = "MCHPRS Service";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/mchprs";
```
Comment (RaitoBezarius) [inline nixos/modules/services/games/mchprs.nix:260]:
```suggestion
        ExecStart = "${lib.getExe cfg.package}";
```

File: pkgs/games/mchprs/default.nix [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    openssl
+    sqlite
+    zlib
+  ] ++ lib.optionals stdenv.isDarwin [
+    darwin.apple_sdk.frameworks.CoreFoundation
+    darwin.apple_sdk.frameworks.Security
+  ];
+
+  meta = with lib; {
+    description = "A multithreaded Minecraft server built for redstone";
```
Comment (RaitoBezarius) [inline pkgs/games/mchprs/default.nix:45]:
```suggestion
    mainProgram = "mchprs";
    description = "A multithreaded Minecraft server built for redstone";
```

File: nixos/modules/services/games/mchprs.nix [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+
+            # Settings were declarative before, no need to back up anything
+            cp -f ${configTomlFile} Config.toml
+
+          else
+
+            # Declarative settings for the first time, backup stateful files
+            cp -b --suffix=.stateful ${configTomlFile} Config.toml
+
+            echo "Autogenerated file that signifies that this server configuration is managed declaratively by NixOS" \
```
Comment (RaitoBezarius) [inline nixos/modules/services/games/mchprs.nix:305]:
```suggestion
            echo "Autogenerated file that implies that this server configuration is managed declaratively by NixOS" \
```

File: nixos/modules/services/games/mchprs.nix [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+            # Whitelist was declarative before, no need to back up anything
+            ln -sf ${whitelistFile} whitelist.json
+
+          else
+
+            # Declarative whitelist for the first time, backup stateful files
+            ln -sb --suffix=.stateful ${whitelistFile} whitelist.json
+
+            echo "Autogenerated file that signifies that this server's whitelist is managed declaratively by NixOS" \
```
Comment (RaitoBezarius) [inline nixos/modules/services/games/mchprs.nix:324]:
```suggestion
            echo "Autogenerated file that implies that this server's whitelist is managed declaratively by NixOS" \
```

File: nixos/modules/services/games/mchprs.nix [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+    };
+
+    networking.firewall = mkIf (cfg.declarativeSettings && cfg.openFirewall) {
+      allowedUDPPorts = [ cfg.settings.port ];
+      allowedTCPPorts = [ cfg.settings.port ];
+    };
+  };
+
+  meta.maintainers = with maintainers; [ gdd ];
+
```
Comment (RaitoBezarius) [inline nixos/modules/services/games/mchprs.nix:342]:
```suggestion
```

File: nixos/modules/services/games/mchprs.nix [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+          https://www.freedesktop.org/software/systemd/man/systemd.time.html#Parsing%20Time%20Spans.
+          If `null`, then the server is not restarted automatically.
+        '';
+      };
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.mchprs;
+        defaultText = literalExpression "pkgs.mchprs";
+        example = literalExpression "pkgs.mchprs";
```
Comment (RaitoBezarius) [inline nixos/modules/services/games/mchprs.nix:80]:
```suggestion
```

File: nixos/modules/services/games/mchprs.nix [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+let
+  cfg = config.services.mchprs;
+  settingsFormat = pkgs.formats.toml { };
+
+  whitelistFile = pkgs.writeText "whitelist.json"
+    (builtins.toJSON
+      (mapAttrsToList (n: v: { name = n; uuid = v; }) cfg.whitelist.list));
+
+  configToml =
+    filterAttrs (n: _: !(elem n [ "address" "port" ])) cfg.settings //
```
Comment (RaitoBezarius) [inline nixos/modules/services/games/mchprs.nix:14]:
```suggestion
    (removeAttrs cfg.settings [ "address" "port" ]) //
```

**Comment by RaitoBezarius** [comment]
Result of `nixpkgs-review pr 240913` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mchprs</li>
  </ul>
</details>

**Pull request merged**